src/libstore/outputs-spec.cc

```diff
@@ -6,57 +6,153 @@
-    static std::regex regex(R"((.*)\^((\*)|([a-z]+(,[a-z]+)*)))");
+    return std::visit(overloaded {
+        [&](const OutputsSpec::All &) {
+            return true;
+        },
+        [&](const OutputsSpec::Names & outputNames) {
+            return outputNames.find(outputName) != outputNames.end();
+        },
+    }, raw());
+}
+
+
+std::optional<OutputsSpec> OutputsSpec::parseOpt(std::string_view s)
+{
+    static std::regex regex(R"((\*)|([a-z]+(,[a-z]+)*))");
 
     std::smatch match;
-    if (!std::regex_match(s, match, regex))
-        return {s, DefaultOutputs()};
+    std::string s2 { s }; // until some improves std::regex
+    if (!std::regex_match(s2, match, regex))
+        return std::nullopt;
+
+    if (match[1].matched)
+        return { OutputsSpec::All {} };
+
+    if (match[2].matched)
+        return { tokenizeString<OutputsSpec::Names>(match[2].str(), ",") };
+
+    assert(false);
+}
 
-    if (match[3].matched)
-        return {match[1], AllOutputs()};
 
-    return {match[1], tokenizeString<OutputNames>(match[4].str(), ",")};
+OutputsSpec OutputsSpec::parse(std::string_view s)
+{
+    std::optional spec = OutputsSpec::parseOpt(s);
+    if (!spec)
+        throw Error("Invalid outputs specifier: '%s'", s);
+    return *spec;
 }
 
+
+std::pair<std::string_view, ExtendedOutputsSpec> ExtendedOutputsSpec::parse(std::string_view s)
+{
+    auto found = s.rfind('^');
+
+    if (found == std::string::npos)
+        return { s, ExtendedOutputsSpec::Default {} };
+
+    auto spec = OutputsSpec::parse(s.substr(found + 1));
+    return { s.substr(0, found), ExtendedOutputsSpec::Explicit { spec } };
+}
+
+
+std::string OutputsSpec::print() const
+{
+    return std::visit(overloaded {
+        [&](const OutputsSpec::All &) -> std::string {
+            return "*";
+        },
+        [&](const OutputsSpec::Names & outputNames) -> std::string {
+            return concatStringsSep(",", outputNames);
+        },
+    }, raw());
+}
+
+
 std::string ExtendedOutputsSpec::print() const
 {
     return std::visit(overloaded {
         [&](const ExtendedOutputsSpec::Default &) -> std::string {
             return "";
         },
-        [&](const ExtendedOutputsSpec::All &) -> std::string {
-            return "*";
+        [&](const ExtendedOutputsSpec::Explicit & outputSpec) -> std::string {
+            return "^" + outputSpec.print();
+        },
+    }, raw());
+}
+
+
+bool OutputsSpec::merge(const OutputsSpec & that)
+{
+    return std::visit(overloaded {
+        [&](OutputsSpec::All &) {
+            /* If we already refer to all outputs, there is nothing to do. */
+            return false;
         },
-        [&](const ExtendedOutputsSpec::Names & outputNames) -> std::string {
-            return "^" + concatStringsSep(",", outputNames);
+        [&](OutputsSpec::Names & theseNames) {
+            return std::visit(overloaded {
+                [&](const OutputsSpec::All &) {
+                    *this = OutputsSpec::All {};
```

Comment:
I wonder if this is strictly legal C++...

Reply:
@tfc? I am not sure.

Reply:
It's not illegal, but I would certainly consider it a dangerous pattern.
Is it maybe possible to refactor this so we have a merge function that is not a member of one of the to-be-merged items?

Reply:
It's gone!